Search with more flexible configuration

francocatena opened the pull request:
We use PostgreSQL full text search a lot and with the previous options we can not take advantage of it.

With this changes we can add <code>:operator</code> for customize the search operator (defaults to 'LIKE'), <code>:mask</code> for customize the query restriction format (defaults to '%%%s%') and <code>:column_operator</code> for customize the function applied to the column (defaults to 'LOWER(%s)')

The place holder %s can be used in the <code>:mask</code> and <code>:column_operator</code> arguments to indicate the string place or the column name respectively

For example, if you want to use PostgreSQL full text search:

<code>Model.auto_complete_for_name(:operator => '@@', :mask => 'to_tsquery(%s)', :column_operator => 'to_tsvector(%s)')</code>

This would result in a query like <code>"SELECT \* FROM models WHERE to_tsvector(name) @@ to_tsquery('some text') ORDER BY name ASC LIMIT 10"</code>

This also is usefull if you want to search only using an implicit % at the end instead of at the beginning and the end.

I think this could be useful to someone else, thank you again for this great plugin, regards.

Franco Catena.

francocatena committed:
Search with more flexible configuration

Now we can add :operator for customize the search operator (defaults to 'LIKE'), :mask for customize the query restriction formar (defaults to '%%%s%') and :column_operator for customize the function applied to the column (defaults to 'LOWER(%s)')

The place holder %s can be used in the :mask and :column_operator arguments to indicate the string place or the column name respectively

For example, if you want to use PostgreSQL full text search:

Model.auto_complete_for_name(:operator => '@@', :mask => 'to_tsquery(%s)', :column_operator => 'to_tsvector(%s)')

This would result in a query like "SELECT * FROM models WHERE to_tsvector(name) @@ to_tsquery('some text') ORDER BY name ASC LIMIT 10"

francocatena committed:
Dependencies update (Rails 3.0.5 and Arel 2.0.9)

francocatena commented:
Do not pull yet, I discover an error. I will come back with a fix =), sorry for that.

Regards.

francocatena committed:
Added a new option :parameter_operator

Usefull when you whant do something like "LIKE LOWER(?)"

grosser commented:
Looks very complicated to me, maybe something simpler like :query=> "%{fields} LIKE LOWER(%{query})" would make these options unnecessary and more readable

francocatena committed:
Changes in flexible search options

Now we can customize the search query with :query => '%{field} LIKE LOWER(%{query})' and the values mask with :mask => '%{value}%'. In this example the generated query is: 'SELECT * FROM models WHERE field_name LIKE LOWER('query_string%') ORDER BY field_name ASC LIMIT 10'

francocatena commented:
You are right, this is way to complex =), I rewrite it with your sugestions, but I can not exclude the mask options to control the values mask. If you find a cleaner way, please tell me.

I am closing this pull request and making another. Regards.

Franco Catena.
